Code Block displays HTML entities

Let me first thank you for the highly anticipated code block feature! Apart from this issue it seems to work great and covers all basics already.

## 📝 Provide detailed reproduction steps (if any)

1. Create empty code block:
![image](https://user-images.githubusercontent.com/617641/70173038-4385b200-16d2-11ea-9c55-c4c1383e06b4.png)

2. Call `textEditor.getData()` and get `<pre><code class="language-text-plain">&nbsp;</code></pre><p>text</p>`

3. Create new instance of CKEditor and call `textEditor.getData('<pre><code class="language-text-plain">&nbsp;</code></pre><p>text</p>')` - i.e. with what `.getData()` previously returned.

### ✔️ Expected result

![image](https://user-images.githubusercontent.com/617641/70173038-4385b200-16d2-11ea-9c55-c4c1383e06b4.png)

### ❌ Actual result

![image](https://user-images.githubusercontent.com/617641/70173197-a5461c00-16d2-11ea-9442-b6b4c8d9f830.png)


## 📃 Other details

Similar problem happens when code block contains "&", after which is returned from `.getData()` as `&amp;` but then displayed literally from `.setData()`.

* Browser: latest chrome/firefox
* OS: Ubuntu
* CKEditor version: 16.0.0
* Installed CKEditor plugins: …

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

I checked and this should work:

```diff
diff --git a/packages/ckeditor5-code-block/src/codeblockediting.js b/packages/ckeditor5-code-block/src/codeblockediting.js
index e819e08933..7e769d8824 100644
--- a/packages/ckeditor5-code-block/src/codeblockediting.js
+++ b/packages/ckeditor5-code-block/src/codeblockediting.js
@@ -128,7 +128,7 @@ export default class CodeBlockEditing extends Plugin {
                   editor.editing.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model, normalizedLanguagesDefs, true ) );
                   editor.data.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model, normalizedLanguagesDefs ) );
                   editor.data.downcastDispatcher.on( 'insert:softBreak', modelToDataViewSoftBreakInsertion( model ), { priority: 'high' } );
-                  editor.data.upcastDispatcher.on( 'element:pre', dataViewToModelCodeBlockInsertion( editor.data, normalizedLanguagesDefs ) );
+                  editor.data.upcastDispatcher.on( 'element:pre', dataViewToModelCodeBlockInsertion( editor.data, editor.editing.view, normalizedLanguagesDefs ) );

                   // Intercept the clipboard input (paste) when the selection is anchored in the code block and force the clipboard
                   // data to be pasted as a single plain text. Otherwise, the code lines will split the code block and
diff --git a/packages/ckeditor5-code-block/src/converters.js b/packages/ckeditor5-code-block/src/converters.js
index 7ee4a6324f..2fdd5b7845 100644
--- a/packages/ckeditor5-code-block/src/converters.js
+++ b/packages/ckeditor5-code-block/src/converters.js
@@ -131,7 +131,7 @@ export function modelToDataViewSoftBreakInsertion( model ) {
  * configuration passed to the feature.
  * @returns {Function} Returns a conversion callback.
  */
-export function dataViewToModelCodeBlockInsertion( dataController, languageDefs ) {
+export function dataViewToModelCodeBlockInsertion( dataController, editingView, languageDefs ) {
          // Language names associated with CSS classes:
          //
          //                {
@@ -183,8 +183,11 @@ export function dataViewToModelCodeBlockInsertion( dataController, languageDefs
                            writer.setAttribute( 'language', defaultLanguageName, codeBlock );
                   }

-                  const stringifiedElement = dataController.processor.toData( viewChild );
-                  const textData = extractDataFromCodeElement( stringifiedElement );
+                  const textData = [ ...editingView.createRangeIn( viewChild ) ]
+                           .filter( current => current.type === 'text' )
+                           .map( ( { item } ) => item.data )
+                           .join( '' );
+
                   const fragment = rawSnippetTextToModelDocumentFragment( writer, textData );

                   writer.append( fragment, codeBlock );
@@ -234,14 +237,3 @@ export function dataViewToModelCodeBlockInsertion( dataController, languageDefs
                   }
          };
 }
-
-// Returns content of `<pre></pre>` with unescaped html inside.
-//
-// @param {String} stringifiedElement
-function extractDataFromCodeElement( stringifiedElement ) {
-         const data = new RegExp( /^<code[^>]*>([\S\s]*)<\/code>$/ ).exec( stringifiedElement )[ 1 ];
-
-         return data
-                  .replace( /&lt;/g, '<' )
-                  .replace( /&gt;/g, '>' );
-}
```

The change causes [two failing tests](https://github.com/ckeditor/ckeditor5/blob/3011e37768225dfe928f3e3321753fc04ca58ff2/packages/ckeditor5-code-block/tests/codeblockediting.js#L892-L909), though. 

But I think that they were wrong in the first place and that's the reason why they fail. If anyone loads `editor.setData( '<pre><code><p>Foo</p></code></pre>' );` it's an invalid HTML.

```
CodeBlockEditing
    data pipeline v -> m conversion
      ✖ should convert pre > code with HTML inside
        Chrome 81.0.4044 (Mac OS X 10.15.4)
      AssertionError: expected '<codeBlock language="plaintext">[]Foo<softBreak></softBreak>Bar</codeBlock>' to equal '<codeBlock language="plaintext">[]<p>Foo</p><softBreak></softBreak><p>Bar</p></codeBlock>'

      + expected - actual

      -<codeBlock language="plaintext">[]Foo<softBreak></softBreak>Bar</codeBlock>
      +<codeBlock language="plaintext">[]<p>Foo</p><softBreak></softBreak><p>Bar</p></codeBlock>

    at Context.eval (webpack:///./packages/ckeditor5-code-block/tests/codeblockediting.js?:913:122)


      ✖ should convert pre > code tag with HTML and nested pre > code tag
        Chrome 81.0.4044 (Mac OS X 10.15.4)
      AssertionError: expected '<codeBlock language="plaintext">[]FooBarBiz</codeBlock>' to equal '<codeBlock language="plaintext">[]<p>Foo</p><pre>Bar</pre><p>Biz</p></codeBlock>'

      + expected - actual

      -<codeBlock language="plaintext">[]FooBarBiz</codeBlock>
      +<codeBlock language="plaintext">[]<p>Foo</p><pre>Bar</pre><p>Biz</p></codeBlock>

    at Context.eval (webpack:///./packages/ckeditor5-code-block/tests/codeblockediting.js?:925:122)
```

--- a/packages/ckeditor5-code-block/src/codeblockediting.js
+++ b/packages/ckeditor5-code-block/src/codeblockediting.js
@@ -128,7 +128,7 @@ export default class CodeBlockEditing extends Plugin {
 		editor.editing.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model, normalizedLanguagesDefs, true ) );
 		editor.data.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model, normalizedLanguagesDefs ) );
 		editor.data.downcastDispatcher.on( 'insert:softBreak', modelToDataViewSoftBreakInsertion( model ), { priority: 'high' } );
-		editor.data.upcastDispatcher.on( 'element:pre', dataViewToModelCodeBlockInsertion( editor.data, normalizedLanguagesDefs ) );
+		editor.data.upcastDispatcher.on( 'element:pre', dataViewToModelCodeBlockInsertion( editor.editing.view, normalizedLanguagesDefs ) );
 
 		// Intercept the clipboard input (paste) when the selection is anchored in the code block and force the clipboard
 		// data to be pasted as a single plain text. Otherwise, the code lines will split the code block and

--- a/packages/ckeditor5-code-block/src/converters.js
+++ b/packages/ckeditor5-code-block/src/converters.js
@@ -126,12 +126,12 @@ export function modelToDataViewSoftBreakInsertion( model ) {
  *
  *		<codeBlock language="javascript">foo();<softBreak></softBreak>bar();</codeBlock>
  *
- * @param {module:engine/controller/datacontroller~DataController} dataController
+ * @param {module:engine/view/view~View} editingView
  * @param {Array.<module:code-block/codeblock~CodeBlockLanguageDefinition>} languageDefs The normalized language
  * configuration passed to the feature.
  * @returns {Function} Returns a conversion callback.
  */
-export function dataViewToModelCodeBlockInsertion( dataController, languageDefs ) {
+export function dataViewToModelCodeBlockInsertion( editingView, languageDefs ) {
 	// Language names associated with CSS classes:
 	//
 	//		{
@@ -183,8 +183,12 @@ export function dataViewToModelCodeBlockInsertion( dataController, languageDefs 
 			writer.setAttribute( 'language', defaultLanguageName, codeBlock );
 		}
 
-		const stringifiedElement = dataController.processor.toData( viewChild );
-		const textData = extractDataFromCodeElement( stringifiedElement );
+		// HTML elements are invalid content for `<code>`.
+		// Read only text nodes.
+		const textData = [ ...editingView.createRangeIn( viewChild ) ]
+			.filter( current => current.type === 'text' )
+			.map( ( { item } ) => item.data )
+			.join( '' );
 		const fragment = rawSnippetTextToModelDocumentFragment( writer, textData );
 
 		writer.append( fragment, codeBlock );
@@ -233,15 +237,4 @@ export function dataViewToModelCodeBlockInsertion( dataController, languageDefs 
 			data.modelCursor = data.modelRange.end;
 		}
 	};
-}
-
-// Returns content of `<pre></pre>` with unescaped html inside.
-//
-// @param {String} stringifiedElement
-function extractDataFromCodeElement( stringifiedElement ) {
-	const data = new RegExp( /^<code[^>]*>([\S\s]*)<\/code>$/ ).exec( stringifiedElement )[ 1 ];
-
-	return data
-		.replace( /&lt;/g, '<' )
-		.replace( /&gt;/g, '>' );
 }

--- a/packages/ckeditor5-code-block/tests/codeblockediting.js
+++ b/packages/ckeditor5-code-block/tests/codeblockediting.js
@@ -889,29 +889,39 @@ describe( 'CodeBlockEditing', () => {
 			);
 		} );
 
+		// Undesired by expected. There is an issue with identifying the correct filler type.
+		// <code> is inline, so dom-to-view converter expects an inline filler.
+		it( 'should convert pre > code with only &nbsp; inside to a codeBlock with &nbsp;', () => {
+			editor.setData( '<pre><code>&nbsp;</code></pre>' );
+
+			expect( getModelData( model ) ).to.equal(
+				'<codeBlock language="plaintext">[]\u00a0</codeBlock>'
+			);
+		} );
+
 		it( 'should convert pre > code with HTML inside', () => {
 			editor.setData( '<pre><code><p>Foo</p>\n<p>Bar</p></code></pre>' );
 
 			expect( getModelData( model ) ).to.equal(
 				'<codeBlock language="plaintext">[]' +
-					'<p>Foo</p>' +
+					'Foo' +
 					'<softBreak></softBreak>' +
-					'<p>Bar</p>' +
+					'Bar' +
 				'</codeBlock>'
 			);
 		} );
 
-		it( 'should convert pre > code tag with HTML and nested pre > code tag', () => {
-			editor.setData( '<pre><code><p>Foo</p><pre>Bar</pre><p>Biz</p></code></pre>' );
+		it( 'should convert pre > code tag with HTML and nested pre > code tag and use only the text content of invalid HTML tags', () => {
+			editor.setData( '<pre><code><p>Foo</p><pre><code>Bar</code></pre><p>Biz</p></code></pre>' );
 
 			expect( getModelData( model ) ).to.equal(
-				'<codeBlock language="plaintext">[]<p>Foo</p><pre>Bar</pre><p>Biz</p></codeBlock>' );
+				'<codeBlock language="plaintext">[]FooBarBiz</codeBlock>' );
 		} );
 
 		it( 'should convert pre > code tag with escaped html content', () => {
-			editor.setData( '<pre><code>&lt;div&gt;&lt;p&gt;Foo&lt;/p&gt;&lt;/div&gt;</code></pre>' );
+			editor.setData( '<pre><code>&lt;div&gt;&lt;p&gt;Foo&apos;s&amp;&quot;bar&quot;&lt;/p&gt;&lt;/div&gt;</code></pre>' );
 
-			expect( getModelData( model ) ).to.equal( '<codeBlock language="plaintext">[]<div><p>Foo</p></div></codeBlock>' );
+			expect( getModelData( model ) ).to.equal( '<codeBlock language="plaintext">[]<div><p>Foo\'s&"bar"</p></div></codeBlock>' );
 		} );
 
 		it( 'should be overridable', () => {

